numpy/core/src/umath/ufunc_object.c

```diff
@@ -2456,7 +2497,7 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
      */
     core_dim_ixs_size = 0;
     for (i = 0; i < nop; ++i) {
-        core_dim_ixs_size += ufunc->core_num_dims[i];
+        core_dim_ixs_size += keepdims && i >= nin ? 1: ufunc->core_num_dims[i];
```

Comment:
Instead of computing this over and over again, why not something like this at the start:
```C
npy_intp *core_num_dims;
if (keepdims) {
    core_num_dims = alloca(sizeof(*core_num_dims) * NPY_MAX_DIMS);
    for (i = 0; i < nin; ++i) {
         core_num_dims[i] = ufunc->core_num_dims[i];
    }
    for(; i < nop; i++) {
        core_num_dims[i] = 1;
    } 
}
else {
    core_num_dims = ufunc->core_num_dims;
}
```

Reply:
OK, better indeed, especially since all dimensions are just 1.

Reply:
It seems appveyor/windows does not like `alloca`, so I now just use a static buffer.